roomserver/internal/input/input_events.go

```diff
@@ -228,10 +302,115 @@ func (r *Inputer) processRoomEvent(
 	return nil
 }
 
+func (r *Inputer) checkForMissingAuthEvents(
+	ctx context.Context,
+	logger *logrus.Entry,
+	event *gomatrixserverlib.HeaderedEvent,
+	auth *gomatrixserverlib.AuthEvents,
+	known map[string]*types.Event,
+	servers []gomatrixserverlib.ServerName,
+) error {
+	authEventIDs := event.AuthEventIDs()
+	if len(authEventIDs) == 0 {
+		return nil
+	}
+
+	unknown := map[string]struct{}{}
+
+	authEvents, err := r.DB.EventsFromIDs(ctx, authEventIDs)
+	if err != nil {
+		return fmt.Errorf("r.DB.EventsFromIDs: %w", err)
+	}
+	for _, event := range authEvents {
```

Comment:
It's not safe to do this because `EventsFromIDs` calls `BulkSelectEventNID` which omits missing events:
```go
// bulkSelectEventNIDs returns a map from string event ID to numeric event ID.
// If an event ID is not in the database then it is omitted from the map.
func (s *eventStatements) BulkSelectEventNID(ctx context.Context, eventIDs []string) (map[string]types.EventNID, error) {
```

Sidenote: we should be really clear on this as it's a massive footgun in other places too, probably having something like `SelectFoo(args...) (result, hasMissing, error)` would be a better pattern.

You need to loop `authEventIDs`.